Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp

```diff
@@ -104,19 +104,7 @@ void WebPermissionController::permissionChanged(WebCore::PermissionName permissi
         if (observer.descriptor().name != permissionName || observer.origin().topOrigin != topOrigin)
             return;
 
-        auto* context = observer.context();
-        if (!context)
-            continue;
-
-        auto source = WebCore::Permissions::sourceFromContext(*context);
-        // FIXME: Add support for workers.
-        ASSERT(source && *source == WebCore::PermissionQuerySource::Window);
-
-        auto* page = downcast<WebCore::Document>(context)->page();
-        if (!page)
-            continue;
-
-        query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, page, *source, [observer = WeakPtr { observer }](auto newState) {
+        query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, observer.page(), observer.source(), [observer = WeakPtr { observer }](auto newState) {
```

Comment:
Can we add an `ASSERT(isMainThread());` at the beginning of this function since the code seems to assume we're on the main thread.